lib/LANraragi/Utils/Database.pm

```diff
@@ -303,4 +304,20 @@ sub invalidate_isnew_cache {
     $redis->quit();
 }
 
+sub save_computed_tagrules {
+    my ( $tagrules ) = @_;
+    my $redis = LANraragi::Model::Config->get_redis;
+    $redis->del("LRR_TAGRULES");
+    $redis->lpush("LRR_TAGRULES", reverse flat(@$tagrules)) if (@$tagrules);
+    $redis->quit();
+    return;
+}
+
+sub get_computed_tagrules {
+    my $redis = LANraragi::Model::Config->get_redis;
+    my @flattened_rules = $redis->lrange("LRR_TAGRULES", 0,-1);
```

Comment:
As you said earlier, this won't work properly if the user hasn't changed the settings, since the tag rules won't be calculated yet.  

This sub should probably include a check like `if ($redis->exists("LRR_TAGRULES"))`, and calculate/save the rules in case the list doesn't exist in Redis yet.

Reply:
I can't believe I forgot it :facepalm: